File: src/reasoners/ssr.py
from __future__ import annotations

import json
import re
from collections import Counter
from dataclasses import dataclass
from typing import List, Optional, Tuple

from .base import Reasoner, ReasoningTrace


@dataclass
class SocraticStep:
    index: int
    question: str
    answer: str
    confidence: float | None = None
    samples: List[str] | None = None


class SSRReasoner(Reasoner):
    """
    Socratic Self-Refinement:
    - initial solution
    - decompose into steps
    - verify each step via sampling
    - refine weakest step
    """

    def solve(self, query: str, initial_solution: Optional[str] = None) -> ReasoningTrace:
        max_iters = self.config.get("max_iterations", 3)
        m_samples = self.config.get("m_samples", 8)
        mode = self.config.get("mode", "plan")

        reasoning, answer = self._initial(query, initial_solution)

        if mode == "plan":
            reasoning, answer = self._refine_plan(query, reasoning, answer)

        history = []
        for it in range(max_iters):
            steps = self._decompose(query, reasoning, answer)
            steps = self._verify(query, steps, m_samples)
            updated = self._refine(query, reasoning, answer, steps)
            if updated is None:
                break
            reasoning, answer = updated
            history.append(
                {
                    "iteration": it + 1,
                    "steps": [{"index": step.index, "conf": step.confidence} for step in steps],
                }
            )

        return ReasoningTrace(reasoning, answer, {"method": "ssr", "history": history})

    def _initial(self, query: str, init: Optional[str]) -> Tuple[str, str]:
        if init:
            return self._split(init)
        resp = self._chat(
            [
                {
                    "role": "system",
                    "content": "Think step by step like a careful mathematician.",
                },
                {
                    "role": "user",
                    "content": query + "\n\nEnd with 'Final answer: ...'.",
                },
            ]
        )
        return self._split(resp)

    def _refine_plan(self, query: str, reasoning: str, answer: str) -> Tuple[str, str]:
        prompt = f"""
[Problem]
{query}

[Draft]
{reasoning}
Final answer: {answer}

1. Summarize a plan.
2. Improve the plan.
3. Rewrite cleanly.

End with 'Final answer: ...'.
"""
        resp = self._chat([{"role": "user", "content": prompt}])
        return self._split(resp)

    def _decompose(self, query: str, reasoning: str, answer: str) -> List[SocraticStep]:
        prompt = f"""
[Problem]
{query}

[Solution]
{reasoning}
Final answer: {answer}

Break this into <= 8 steps. Return JSON array only.
"""
        raw = self._chat(
            [{"role": "user", "content": prompt}],
            temperature=0.0,
            max_tokens=512,
        )
        try:
            data = json.loads(raw.strip("` \n"))
            return [
                SocraticStep(
                    index=int(step.get("index", i + 1)),
                    question=str(step["question"]),
                    answer=str(step["answer"]),
                )
                for i, step in enumerate(data)
            ]
        except Exception:
            return [
                SocraticStep(1, "Whole reasoning", reasoning + f"\nFinal answer: {answer}")
            ]

    def _verify(self, query: str, steps: List[SocraticStep], m: int) -> List[SocraticStep]:
        for idx, step in enumerate(steps):
            prev = "\n".join(f"Step {s.index}: {s.answer}" for s in steps[:idx])
            samples: list[str] = []
            for _ in range(m):
                prompt = f"""
[Problem]
{query}

Previous steps:
{prev or "(none)"}

Re-solve the sub-question:
{step.question}

Short answer only.
"""
                samples.append(
                    self._chat(
                        [{"role": "user", "content": prompt}],
                        temperature=0.8,
                        max_tokens=64,
                    )
                )
            step.samples = samples
            gold = self._norm(step.answer)
            matches = sum(1 for sample in samples if self._norm(sample) == gold)
            step.confidence = matches / m
        return steps

    def _refine(
        self,
        query: str,
        reasoning: str,
        answer: str,
        steps: List[SocraticStep],
    ) -> tuple[str, str] | None:
        bad = min(steps, key=lambda step: step.confidence)
        if bad.confidence is None or bad.confidence > 0.8:
            return None

        counts = Counter(self._norm(sample) for sample in (bad.samples or []))
        best_norm, _ = counts.most_common(1)[0]

        if best_norm == self._norm(bad.answer):
            return None

        best_raw = next(
            (sample for sample in (bad.samples or []) if self._norm(sample) == best_norm),
            bad.answer,
        )

        prompt = f"""
[Problem]
{query}

[Old solution]
{reasoning}
Final answer: {answer}

Bad step:
Q: {bad.question}
Old A: {bad.answer}
Better A: {best_raw}

Rewrite full solution consistently.
End with 'Final answer: ...'.
"""
        out = self._chat([{"role": "user", "content": prompt}], temperature=0.7)
        return self._split(out)

    @staticmethod
    def _norm(s: str) -> str:
        s = s.strip().lower()
        s = re.sub(r"[ \t\n\r]+", " ", s)
        return s.rstrip(".,;:!?")


__all__ = ["SocraticStep", "SSRReasoner"]

File: tui/packages/agent-capabilities/src/reasoning/ssr.ts
// ── SSR: Socratic Self-Refinement ──────────────────────────────────────────────
// Adapted from Python src/reasoners/ssr.py.
//
// Flow: initial solution → decompose into steps → verify each step via sampling
// → refine weakest step → repeat.

import { BaseReasoner, type ReasoningTrace } from "./base.js";

export interface SocraticStep {
	index: number;
	question: string;
	answer: string;
	confidence: number | null;
	samples: string[] | null;
}

interface SSRConfig {
	maxIterations?: number;
	mSamples?: number;
	mode?: "plan" | "direct";
	temperature?: number;
	maxTokens?: number;
	[key: string]: unknown;
}

export class SSRReasoner extends BaseReasoner {
	config: SSRConfig;

	constructor(
		llm: import("@logician/agent-core/agent/backend.ts").LLMBackend,
		config: SSRConfig = {},
	) {
		super(llm, config);
		this.config = config;
	}

	async solve(
		query: string,
		initialSolution?: string | undefined | null,
	): Promise<ReasoningTrace> {
		const maxIter = this.config.maxIterations ?? 3;
		const mSamples = this.config.mSamples ?? 8;
		const mode = this.config.mode ?? "plan";

		let [reasoning, answer] = await this._initial(query, initialSolution);

		if (mode === "plan") {
			[reasoning, answer] = await this._refinePlan(query, reasoning, answer);
		}

		const history: {
			iteration: number;
			steps: { index: number; conf: number | null }[];
		}[] = [];

		for (let it = 0; it < maxIter; it++) {
			const steps = await this._decompose(query, reasoning, answer);
			const verified = await this._verify(query, steps, mSamples);
			const updated = await this._refine(query, reasoning, answer, verified);
			if (updated === null) break;
			[reasoning, answer] = updated;
			history.push({
				iteration: it + 1,
				steps: verified.map((s) => ({ index: s.index, conf: s.confidence })),
			});
		}

		return { reasoning, answer, metadata: { method: "ssr", history } };
	}

	private async _initial(
		query: string,
		init: string | undefined | null,
	): Promise<[string, string]> {
		if (init) return this._split(init);
		const resp = await this._chat([
			{
				role: "system",
				content: "Think step by step like a careful mathematician.",
			},
			{ role: "user", content: `${query}\n\nEnd with 'Final answer: ...'.` },
		]);
		return this._split(resp);
	}

	private async _refinePlan(
		query: string,
		reasoning: string,
		answer: string,
	): Promise<[string, string]> {
		const prompt = [
			"[Problem]",
			query,
			"[Draft]",
			reasoning,
			"Final answer:",
			answer,
			"",
			"1. Summarize a plan.",
			"2. Improve the plan.",
			"3. Rewrite cleanly.",
			"End with 'Final answer: ...'.",
		].join("\n");
		const resp = await this._chat([{ role: "user", content: prompt }]);
		return this._split(resp);
	}

	private async _decompose(
		query: string,
		reasoning: string,
		answer: string,
	): Promise<SocraticStep[]> {
		const prompt = [
			"[Problem]",
			query,
			"[Solution]",
			reasoning,
			"Final answer:",
			answer,
			"",
			"Break this into <= 8 steps. Return JSON array only.",
		].join("\n");
		const raw = await this._chat([{ role: "user", content: prompt }], {
			temperature: 0.0,
			maxTokens: 512,
		});
		try {
			const cleaned = raw
				.replace(/^`+\s*/, "")
				.replace(/\s*`+$/, "")
				.trim();
			const data = JSON.parse(cleaned) as Array<Record<string, unknown>>;
			return data.map(
				(step, i) =>
					({
						index: Number(step.index ?? i + 1),
						question: String(step.question ?? ""),
						answer: String(step.answer ?? ""),
						confidence: null,
						samples: null,
					}) as SocraticStep,
			);
		} catch (_e: unknown) {
			return [
				{
					index: 1,
					question: "Whole reasoning",
					answer: `${reasoning}\nFinal answer: ${answer}`,
					confidence: null,
					samples: null,
				},
			] as SocraticStep[];
		}
	}

	private async _verify(
		query: string,
		steps: SocraticStep[],
		m: number,
	): Promise<SocraticStep[]> {
		const results: SocraticStep[] = [];
		for (let idx = 0; idx < steps.length; idx++) {
			const step = steps[idx];
			const prev = steps
				.slice(0, idx)
				.map((s) => `Step ${s.index}: ${s.answer}`)
				.join("\n");
			const samples: string[] = [];
			for (let s = 0; s < m; s++) {
				const sample = await this._chat(
					[
						{
							role: "user",
							content: `Problem: ${query}\nPrevious steps:\n${prev || "(none)"}\nRe-solve the sub-question:\n${step.question}\nShort answer only.`,
						},
					],
					{ temperature: 0.8, maxTokens: 64 },
				);
				samples.push(sample);
			}
			const norm = SSRReasoner._norm(step.answer);
			const matches = samples.filter(
				(s) => SSRReasoner._norm(s) === norm,
			).length;
			step.samples = samples;
			step.confidence = matches / m;
			results.push(step);
		}
		return results;
	}

	private async _refine(
		query: string,
		reasoning: string,
		answer: string,
		steps: SocraticStep[],
	): Promise<[string, string] | null> {
		const bad = steps.reduce((worst, s) =>
			(s.confidence ?? 1) < (worst.confidence ?? 1) ? s : worst,
		);
		if (bad.confidence === null || bad.confidence > 0.8) return null;

		const counts = new Map<string, number>();
		for (const sample of bad.samples ?? []) {
			const n = SSRReasoner._norm(sample);
			counts.set(n, (counts.get(n) ?? 0) + 1);
		}
		const [bestNorm] = [...counts.entries()].sort((a, b) => b[1] - a[1])[0];
		if (bestNorm === SSRReasoner._norm(bad.answer)) return null;

		const bestRaw =
			bad.samples?.find((s) => SSRReasoner._norm(s) === bestNorm) ?? bad.answer;

		const prompt = [
			"[Problem]",
			query,
			"[Old solution]",
			reasoning,
			"Final answer:",
			answer,
			"",
			"Bad step:",
			`Q: ${bad.question}`,
			`Old A: ${bad.answer}`,
			`Better A: ${bestRaw}`,
			"",
			"Rewrite full solution consistently.",
			"End with 'Final answer: ...'.",
		].join("\n");
		const out = await this._chat([{ role: "user", content: prompt }], {
			temperature: 0.7,
		});
		return this._split(out);
	}

	private static _norm(s: string): string {
		return s
			.trim()
			.toLowerCase()
			.replace(/[ \t\n\r]+/g, " ")
			.replace(/[.,;:!?]$/, "");
	}
}
